### DISCUSSION TITLE
Parallel performance

### DISCUSSION THREAD

**Discussion by adrianjhpc** [CGNS/CGNS#707] [CGNStalk]
Is there any way to optimise performance for larger files. I'm currently reading a ~800MB file into 96 processes using `cgp_` functionality (setup `CGP_INDEPENDENT` because there are varying numbers of zones per process) and it's taking minutes to get the data. There are ~2000 zones in the file, but it's not large amounts of data. 

If I use our MPI-I/O process to read the same data from a binary file rather than a CGNS file I can do it in seconds.

**Comment by mennodeij** [comment]
1) is the file written with HDF5 format? If not, you can convert it and try again.
2) is the HDF5 library compiled with parallel support? If not, please ensure that it is.
3) the `CGP_INDEPENDENT` use may or may not slow things down. Have you tried with `CGP_COLLECTIVE` to see if read speeds improve?

**Comment by adrianjhpc** [comment]
Thanks for the follow up:
1. Yes, the parallel functionality won't work with ADF anyway.
2. Yes, parallel support is in hdf5.
3. I've tried both `CGP_INDEPENDENT` and `CGP_COLLECTIVE`, they different by 2 seconds overall a total of ~2500 seconds.  This is for reading a 1.5GB file using 244 processes.

**Comment by brtnfld** [comment]
What version of CGNS are you using? Would you happen to know what function is taking the time? Is it the actual read call? Is this on a Lustre or GPFS parallel filesystem? If Lustre, what are your stripe size and count? Does your system have the Darshan logs available?
https://www.mcs.anl.gov/research/projects/darshan/

**Comment by adrianjhpc** [comment]
This is the git development version of CGNS, but I see the same with version 4.3.0. 

Really the only function I'm using is:
```
      call cgp_coord_read_data_f(fid,basenum,zonenum,xnum,
     &     minrange,maxrange,xyz(:,:,:,1),ierr)
```
Three times per zone. There are 6588 zones, but they are splits across the 244 processes in this case  (so 27 zones per process).

This is Lustre, with 1MB stripes and striping across all OSTs, although either way the MPI I/O performance is seconds regardless of striping.

We don't have Darshan available by default, but I can run with Darshan if that would be useful.

**Comment by brtnfld** [comment]
Darshan logs would help determine the issue. In the past cgp_open was slow, so it would be good to eliminate the other cgp functions being the issue.

**Comment by adrianjhpc** [comment]
Running through Darshan and looking at some detailed logs, I can only see `H5D` function calls, not `H5P`. i.e.:
```
H5D     -1      23130105024425992       H5D_OPENS       1708    /lustre/home/nx01/nx01/adrianj/Turbine/eCSE3/G4_rot_ned/mb6588_medium_rotor_inc_fnames.cgns.new:/Base/Block_s1008/ZoneGridConnectivity/bc_3/PointRangeDonor data        /lust
re      lustre
H5D     -1      23130105024425992       H5D_READS       732     /lustre/home/nx01/nx01/adrianj/Turbine/eCSE3/G4_rot_ned/mb6588_medium_rotor_inc_fnames.cgns.new:/Base/Block_s1008/ZoneGridConnectivity/bc_3/PointRangeDonor data        /lust
re      lustre
H5D     -1      23130105024425992       H5D_WRITES      0       /lustre/home/nx01/nx01/adrianj/Turbine/eCSE3/G4_rot_ned/mb6588_medium_rotor_inc_fnames.cgns.new:/Base/Block_s1008/ZoneGridConnectivity/bc_3/PointRangeDonor data        /lust
re      lustre
H5D     -1      23130105024425992       H5D_FLUSHES     0       /lustre/home/nx01/nx01/adrianj/Turbine/eCSE3/G4_rot_ned/mb6588_medium_rotor_inc_fnames.cgns.new:/Base/Block_s1008/ZoneGridConnectivity/bc_3/PointRangeDonor data        /lust
re      lustre
H5D     -1      23130105024425992       H5D_BYTES_READ  35136   /lustre/home/nx01/nx01/adrianj/Turbine/eCSE3/G4_rot_ned/mb6588_medium_rotor_inc_fnames.cgns.new:/Base/Block_s1008/ZoneGridConnectivity/bc_3/PointRangeDonor data        /lust
re      lustre
H5D     -1      23130105024425992       H5D_BYTES_WRITTEN       0       /lustre/home/nx01/nx01/adrianj/Turbine/eCSE3/G4_rot_ned/mb6588_medium_rotor_inc_fnames.cgns.new:/Base/Block_s1008/ZoneGridConnectivity/bc_3/PointRangeDonor data    /lustre  lustre
H5D     -1      23130105024425992       H5D_RW_SWITCHES 0       /lustre/home/nx01/nx01/adrianj/Turbine/eCSE3/G4_rot_ned/mb6588_medium_rotor_inc_fnames.cgns.new:/Base/Block_s1008/ZoneGridConnectivity/bc_3/PointRangeDonor data        /lust
re      lustre
H5D     -1      23130105024425992       H5D_REGULAR_HYPERSLAB_SELECTS   732     /lustre/home/nx01/nx01/adrianj/Turbine/eCSE3/G4_rot_ned/mb6588_medium_rotor_inc_fnames.cgns.new:/Base/Block_s1008/ZoneGridConnectivity/bc_3/PointRangeDonor d
ata     /lustre lustre
H5D     -1      23130105024425992       H5D_IRREGULAR_HYPERSLAB_SELECTS 0       /lustre/home/nx01/nx01/adrianj/Turbine/eCSE3/G4_rot_ned/mb6588_medium_rotor_inc_fnames.cgns.new:/Base/Block_s1008/ZoneGridConnectivity/bc_3/PointRangeDonor d
```

Is this to be expected?

**Comment by adrianjhpc** [comment]
Turns out, I needed to step I_MPI_EXTRA_FILESYSTEM=1 for the Intel MPI library to get sensible performance.
